src/main/java/newbank/server/commands/Command.java

```diff
@@ -1,27 +1,94 @@
 package newbank.server.commands;
 
 import newbank.server.CustomerID;
-import newbank.server.exceptions.CommandInvalidSyntaxException;
-import newbank.server.exceptions.RequestNotAllowedException;
+import newbank.server.NewBank;
 
-/** Abstract representation of a command. */
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Abstract representation of a command.
+ */
 public abstract class Command {
-  public abstract String execute() throws CommandInvalidSyntaxException;
+  protected final NewBank bank;
+  protected final String[] tokens;
+  protected final CustomerID customerID;
+  protected List<Supplier<Optional<String>>> responsibilityChain;
+
+  protected Command(NewBank bank, String[] tokens, CustomerID customerID) {
+    this.bank = bank;
+    this.tokens = tokens;
+    this.customerID = customerID;
+  }
+
+  public abstract String execute();
 
   /**
    * @return the command syntax
    */
-  public String getSyntax() {
-    return "";
+  public abstract String getSyntax();
+
+  /**
+   * Apply each of the functions in the responsibility chain. If any of them
+   * return a String (containing a failure message), stop processing.
+   *
+   * @return `Optional`ly either a string containing a failure message, or null.
+   */
+  protected Optional<String> applyResponsibilityChain() {
+    return responsibilityChain.stream()
+        .map(Supplier::get)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .findFirst();
+  }
+
+  protected Optional<String> requestingHelp() {
+    if (tokens.length >= 2 && tokens[1].equalsIgnoreCase("HELP"))
+      return Optional.of(String.format("SUCCESS: Usage: %s", getSyntax()));
+    return Optional.empty();
   }
 
-  protected void checkLoggedIn(CustomerID customer) throws RequestNotAllowedException {
-    if (!isLoggedIn(customer)) {
-      throw new RequestNotAllowedException();
+  protected Optional<String> mustLogIn() {
+    if (isLoggedIn()) return Optional.empty();
+    return Optional.of("FAIL: Request not allowed, please log in first.");
+  }
+
+  protected Optional<String> incorrectUsage() {
+    String[] expectedTokens = getSyntax().split("\\s");
+    if (tokens.length > expectedTokens.length || incorrectOptionalLiterals(expectedTokens))
```

Comment:
What happens if a command handles exactly 3 token, but only 2 are passed by the user? Is it handled on the `command.execute` level?

Reply:
Hmm, this is handled by `incorrectOptionalLiterals`. I'll add a comment to make it clearer how it does this. It's a bit arcane.